src/Service/WorkloadReportService.php

```diff
@@ -0,0 +1,163 @@
+    }
+
+    /**
+     * Get the workload data based on the specified periods, date calculation method, readable period representation method
+     * and view mode.
+     *
+     * @param array $periods The list of periods
+     * @param callable $getDatesOfPeriod The callable to get the first and last date of a given period
+     * @param callable $getReadablePeriod The callable to get a readable representation of a given period
+     * @param string $viewMode The view mode
+     *
+     * @return WorkloadReportData The workload report data
+     *
+     * @throws \Exception When the calculated roundedLoggedPercentage is null
+     */
+    private function getWorkloadData(array $periods, callable $getDatesOfPeriod, callable $getReadablePeriod, int $currentPeriodNumeric, string $viewMode): WorkloadReportData
+    {
+        $workloadReportData = new WorkloadReportData($viewMode);
+        $workers = $this->workerRepository->findAll();
+
+        foreach ($periods as $period) {
+            // Get period specific readable period representation for table headers.
+            $readablePeriod = $getReadablePeriod($period);
+            $workloadReportData->period->set($period, $readablePeriod);
+        }
+        foreach ($workers as $worker) {
+            $workloadReportWorker = new WorkloadReportWorker();
+            $workloadReportWorker->setEmail($worker->getUserIdentifier());
+            $workloadReportWorker->setWorkload($worker->getWorkload());
+
+            foreach ($periods as $period) {
+                // Add current period match-point (current week-number, month-number etc.)
+                if ($period === $currentPeriodNumeric) {
+                    $workloadReportData->setCurrentPeriodNumeric($period);
+                }
+                // Get first and last date in period.
+                $firstAndLastDate = $getDatesOfPeriod($period);
+
+                // Get all worklogs between the two dates.
+                $worklogs = $this->worklogRepository->findWorklogsByWorkerAndDateRange($worker->getUserIdentifier(), $firstAndLastDate['first'], $firstAndLastDate['last']);
+
+                // Tally up logged hours in gathered worklogs for current period.
+                $loggedHours = 0;
+                foreach ($worklogs as $worklog) {
+                    $loggedHours += ($worklog->getTimeSpentSeconds() / 60 / 60);
+                }
+
+                $workerWorkload = $worker->getWorkload();
+
+                if (!$workerWorkload) {
+                    $workerId = $worker->getUserIdentifier();
+                    throw new \Exception("Workload of worker: $workerId cannot be unset when generating workload report.");
+                }
+
+                // Get total logged percentage based on weekly workload.
+                $roundedLoggedPercentage = $this->getRoundedLoggedPercentage($loggedHours, $workerWorkload, $viewMode);
+
+                if (!$roundedLoggedPercentage) {
```

Comment:
the function getRoundedLoggedPercentage cannot return null according to its return type.